### ISSUE TITLE
DNIe 3.0 (spanish card) does not work

### ISSUE THREAD

**Issue by ocurero** [OpenSC/OpenSC#1246]
### Problem Description

DNIe 3.0 was hit by [CVE-2017-15361 ](https://cve.mitre.org/cgi-bin/cvename.cgi?name=CVE-2017-15361) and the certificates were revoked for all newer cards. I went to the police office to renew the certificates with the new ones that are not revoked. Before the security issue the card worked fine using opensc version 0.17.

With the new certificates, the card doesn't work. Firefox doesn't do anything and pkcs11-tool gives an error.

### Steps to reproduce

1. Use an DNIe 3.0 card.
2. Run command: pkcs11-tool --login --test

### Logs

```
nolina:~ # pkcs11-tool --login --test
Using slot 0 with a present token (0x0)
Logging in to "PIN1 (DNI electrónico)".
Please enter User PIN: 
C_SeedRandom() and C_GenerateRandom():
  seeding (C_SeedRandom) not supported
  ERR: C_GenerateRandom(buf1,0) failed: CKR_ARGUMENTS_BAD (0x7)
Digests:
  all 4 digest functions seem to work
  MD5: OK
  SHA-1: OK
  RIPEMD160: OK
Signatures (currently only for RSA)
  testing key 0 (KprivAutenticacion) 
error: PKCS11 function C_SignFinal failed: rv = CKR_FUNCTION_NOT_SUPPORTED (0x54)
```
[Full log with debug=3](https://gist.githubusercontent.com/ocurero/69fa077a2957fe9a08eccd448ab98972/raw/9b11326bb7cd4adc201b1672e7821818887145cd/opensc-debug.log) (a lot of personal data has been removed)

Thanks!

**Comment by martinpaljak** [comment]
In Estonia the "workaround" was to use ECC keys. Maybe something similar was made in Spain? Or were the keys increased to 3k RSA?

**Comment by frankmorgner** [comment]
Jumping through the log, I found the following:
```
0x7fcbf6f59740 13:04:20.253 [opensc-pkcs11] misc.c:280:session_get_operation: called
0x7fcbf6f59740 13:04:20.253 [opensc-pkcs11] mechanism.c:444:sc_pkcs11_signature_final: called
0x7fcbf6f59740 13:04:20.253 [opensc-pkcs11] mechanism.c:446:sc_pkcs11_signature_final: data length 35
0x7fcbf6f59740 13:04:20.253 [opensc-pkcs11] mechanism.c:459:sc_pkcs11_signature_final: 35 bytes to sign
0x7fcbf6f59740 13:04:20.253 [opensc-pkcs11] framework-pkcs15.c:3663:pkcs15_prkey_sign: Initiating signing operation, mechanism 0x1.
0x7fcbf6f59740 13:04:20.253 [opensc-pkcs11] card.c:407:sc_lock: called
0x7fcbf6f59740 13:04:20.253 [opensc-pkcs11] reader-pcsc.c:612:pcsc_lock: called
0x7fcbf6f59740 13:04:20.253 [opensc-pkcs11] card.c:449:sc_lock: returning with: 0 (Success)
0x7fcbf6f59740 13:04:20.253 [opensc-pkcs11] framework-pkcs15.c:3726:pkcs15_prkey_sign: Selected flags 12. Now computing signature for 35 bytes. 1024 bytes reserved.
0x7fcbf6f59740 13:04:20.253 [opensc-pkcs11] pkcs15-sec.c:320:sc_pkcs15_compute_signature: called
0x7fcbf6f59740 13:04:20.253 [opensc-pkcs11] pkcs15-sec.c:154:format_senv: Card does not support RSA with key length 1920
0x7fcbf6f59740 13:04:20.253 [opensc-pkcs11] pkcs15-sec.c:155:format_senv: returning with: -1408 (Not supported)
0x7fcbf6f59740 13:04:20.253 [opensc-pkcs11] pkcs15-sec.c:327:sc_pkcs15_compute_signature: Could not initialize security environment: -1408 (Not supported)
```
If I remember correctly, RSA keys with 1920 bits aren't affected as well. @ocurero, please try https://github.com/OpenSC/OpenSC/pull/1247 if it fixes your issue.

**Comment by frankmorgner** [comment]
Uhm, I think you're right it needs to be above 3k bits. Anyway, I hope the new card doesn't have this vulnerability.

**Comment by alvaroaalvarez** [comment]
I have same issue, tried pull #1247 and get similar error

```
[alvaro@dell8700 opensc]$ bin/pkcs11-tool -t -l
Using slot 0 with a present token (0x0)
Logging in to "PIN1 (DNI electrónico)".
Please enter User PIN: 
C_SeedRandom() and C_GenerateRandom():
  seeding (C_SeedRandom) not supported
  ERR: C_GenerateRandom(buf1,0) failed: CKR_ARGUMENTS_BAD (0x7)
Digests:
  all 4 digest functions seem to work
  MD5: OK
  SHA-1: OK
  RIPEMD160: OK
Signatures (currently only for RSA)
  testing key 0 (KprivAutenticacion) 
  all 4 signature functions seem to work
error: PKCS11 function C_Sign failed: rv = CKR_ARGUMENTS_BAD (0x7)
Aborting.
```

**Comment by frankmorgner** [comment]
@alvaroaalvarez if you don't install OpenSC, then `bin/pkcs11-tool` will use the global library, which may not contain the fix.

If you're sure that the correct library is used and the error persists, please run `OPENSC_DEBUG=3 pkcs11-tool -t -l` and remove your PIN from the log.

**Comment by alvaroaalvarez** [comment]
I did make install, but to the /opt directory,
It's using the fixed lib because now the error is different.

Error before:
```
0x7f1ccce12740 21:53:56.180 [opensc-pkcs11] framework-pkcs15.c:3726:pkcs15_prkey_sign: Selected flags 12. Now computing signature for 35 bytes. 1024 bytes reserved.
0x7f1ccce12740 21:53:56.180 [opensc-pkcs11] pkcs15-sec.c:320:sc_pkcs15_compute_signature: called
0x7f1ccce12740 21:53:56.180 [opensc-pkcs11] pkcs15-sec.c:154:format_senv: Card does not support RSA with key length 1920
0x7f1ccce12740 21:53:56.180 [opensc-pkcs11] pkcs15-sec.c:155:format_senv: returning with: -1408 (Not supported)
0x7f1ccce12740 21:53:56.180 [opensc-pkcs11] pkcs15-sec.c:327:sc_pkcs15_compute_signature: Could not initialize security environment: -1408 (Not supported)
0x7f1ccce12740 21:53:56.180 [opensc-pkcs11] card.c:459:sc_unlock: called
0x7f1ccce12740 21:53:56.180 [opensc-pkcs11] reader-pcsc.c:662:pcsc_unlock: called
0x7f1ccce12740 21:53:56.184 [opensc-pkcs11] framework-pkcs15.c:3743:pkcs15_prkey_sign: Sign complete. Result -1408.
0x7f1ccce12740 21:53:56.184 [opensc-pkcs11] misc.c:61:sc_to_cryptoki_error_common: libopensc return value: -1408 (Not supported)
0x7f1ccce12740 21:53:56.184 [opensc-pkcs11] mechanism.c:462:sc_pkcs11_signature_final: returning with: 84
0x7f1ccce12740 21:53:56.184 [opensc-pkcs11] mechanism.c:327:sc_pkcs11_sign_final: returning with: 84
0x7f1ccce12740 21:53:56.184 [opensc-pkcs11] pkcs11-object.c:765:C_SignFinal: C_SignFinal() = CKR_FUNCTION_NOT_SUPPORTED
0x7f1ccce12740 21:53:56.184 [opensc-pkcs11] pkcs11-global.c:311:C_Finalize: C_Finalize()
0x7f1ccce12740 21:53:56.184 [opensc-pkcs11] ctx.c:845:sc_cancel: called
```

Now, log file is bigger, so I am guessing the fix did corrected something.
New error:
```
0x7fa0755b7740 21:02:06.963 [opensc-pkcs11] cwa14890.c:1822:cwa_decode_response: returning with: 0 (Success)
0x7fa0755b7740 21:02:06.963 [opensc-pkcs11] card-dnie.c:776:dnie_sm_free_wrapped_apdu: returning with: 0 (Success)
0x7fa0755b7740 21:02:06.963 [opensc-pkcs11] sm.c:173:sc_sm_single_transmit: returning with: 0 (Success)
0x7fa0755b7740 21:02:06.963 [opensc-pkcs11] apdu.c:382:sc_single_transmit: returning with: 0 (Success)
0x7fa0755b7740 21:02:06.963 [opensc-pkcs11] apdu.c:543:sc_transmit: returning with: 0 (Success)
0x7fa0755b7740 21:02:06.963 [opensc-pkcs11] card.c:459:sc_unlock: called
0x7fa0755b7740 21:02:06.963 [opensc-pkcs11] card-dnie.c:1816:dnie_check_sw: called
0x7fa0755b7740 21:02:06.963 [opensc-pkcs11] card-dnie.c:1828:dnie_check_sw: returning with: 0 (Success)
0x7fa0755b7740 21:02:06.963 [opensc-pkcs11] card-dnie.c:1559:dnie_set_security_env: returning with: 0 (Success)
0x7fa0755b7740 21:02:06.963 [opensc-pkcs11] sec.c:78:sc_set_security_env: returning with: 0 (Success)
0x7fa0755b7740 21:02:06.963 [opensc-pkcs11] sec.c:58:sc_compute_signature: called
0x7fa0755b7740 21:02:06.963 [opensc-pkcs11] card-dnie.c:1669:dnie_compute_signature: called
0x7fa0755b7740 21:02:06.963 [opensc-pkcs11] card-dnie.c:1676:dnie_compute_signature: returning with: -1300 (Invalid arguments)
0x7fa0755b7740 21:02:06.964 [opensc-pkcs11] sec.c:62:sc_compute_signature: returning with: -1300 (Invalid arguments)
0x7fa0755b7740 21:02:06.964 [opensc-pkcs11] card.c:459:sc_unlock: called
0x7fa0755b7740 21:02:06.964 [opensc-pkcs11] pkcs15-sec.c:454:sc_pkcs15_compute_signature: use_key() failed: -1300 (Invalid arguments)
0x7fa0755b7740 21:02:06.964 [opensc-pkcs11] card.c:459:sc_unlock: called
0x7fa0755b7740 21:02:06.964 [opensc-pkcs11] reader-pcsc.c:662:pcsc_unlock: called
0x7fa0755b7740 21:02:06.971 [opensc-pkcs11] framework-pkcs15.c:3743:pkcs15_prkey_sign: Sign complete. Result -1300.
0x7fa0755b7740 21:02:06.971 [opensc-pkcs11] misc.c:61:sc_to_cryptoki_error_common: libopensc return value: -1300 (Invalid arguments)
0x7fa0755b7740 21:02:06.971 [opensc-pkcs11] mechanism.c:462:sc_pkcs11_signature_final: returning with: 7
0x7fa0755b7740 21:02:06.971 [opensc-pkcs11] mechanism.c:327:sc_pkcs11_sign_final: returning with: 7
0x7fa0755b7740 21:02:06.971 [opensc-pkcs11] pkcs11-object.c:701:C_Sign: C_Sign() = CKR_ARGUMENTS_BAD
0x7fa0755b7740 21:02:06.971 [opensc-pkcs11] pkcs11-global.c:311:C_Finalize: C_Finalize()
0x7fa0755b7740 21:02:06.971 [opensc-pkcs11] ctx.c:845:sc_cancel: called
0x7fa0755b7740 21:02:06.971 [opensc-pkcs11] reader-pcsc.c:712:pcsc_cancel: called
0x7fa0755b7740 21:02:06.971 [opensc-pkcs11] slot.c:195:card_removed: Generic USB2.0-CRW [Smart Card Reader Interface] (20070818000000000) 00 00: card removed
0x7fa0755b7740 21:02:06.971 [opensc-pkcs11] slot.c:476:slot_token_removed: slot_token_removed(0x0)
0x7fa0755b7740 21:02:06.971 [opensc-pkcs11] pkcs11-session.c:140:sc_pkcs11_close_all_sessions: real C_CloseAllSessions(0x0) 1
0x7fa0755b7740 21:02:06.971 [opensc-pkcs11] pkcs11-session.c:109:sc_pkcs11_close_session: real C_CloseSession(0x21d5b60)
0x7fa0755b7740 21:02:06.971 [opensc-pkcs11] pkcs15-pin.c:826:sc_pkcs15_pincache_clear: called
0x7fa0755b7740 21:02:06.971 [opensc-pkcs11] card-dnie.c:1415:dnie_logout: called
...
0x7fa0755b7740 21:02:06.974 [opensc-pkcs11] card.c:459:sc_unlock: called
0x7fa0755b7740 21:02:06.974 [opensc-pkcs11] reader-pcsc.c:662:pcsc_unlock: called
0x7fa0755b7740 21:02:06.979 [opensc-pkcs11] card-dnie.c:1816:dnie_check_sw: called
0x7fa0755b7740 21:02:06.979 [opensc-pkcs11] card-dnie.c:1821:dnie_check_sw: Expected SM Data Object missing
0x7fa0755b7740 21:02:06.979 [opensc-pkcs11] card-dnie.c:1191:dnie_compose_and_send_apdu: SelectFile() check_sw failed: -1600 (Generic Secure Messaging error)
0x7fa0755b7740 21:02:06.979 [opensc-pkcs11] card-dnie.c:1430:dnie_logout: returning with: 0 (Success)
0x7fa0755b7740 21:02:06.979 [opensc-pkcs11] framework-pkcs15.c:1472:pkcs15_release_token: pkcs15_release_token() not implemented
0x7fa0755b7740 21:02:06.979 [opensc-pkcs11] slot.c:476:slot_token_removed: slot_token_removed(0x1)
0x7fa0755b7740 21:02:06.980 [opensc-pkcs11] pkcs11-session.c:140:sc_pkcs11_close_all_sessions: real C_CloseAllSessions(0x1) 0
0x7fa0755b7740 21:02:06.980 [opensc-pkcs11] slot.c:476:slot_token_removed: slot_token_removed(0x2)
0x7fa0755b7740 21:02:06.980 [opensc-pkcs11] pkcs11-session.c:140:sc_pkcs11_close_all_sessions: real C_CloseAllSessions(0x2) 0
0x7fa0755b7740 21:02:06.980 [opensc-pkcs11] slot.c:476:slot_token_removed: slot_token_removed(0x3)
0x7fa0755b7740 21:02:06.980 [opensc-pkcs11] pkcs11-session.c:140:sc_pkcs11_close_all_sessions: real C_CloseAllSessions(0x3) 0
0x7fa0755b7740 21:02:06.980 [opensc-pkcs11] pkcs15.c:1273:sc_pkcs15_unbind: called
0x7fa0755b7740 21:02:06.980 [opensc-pkcs11] pkcs15-pin.c:826:sc_pkcs15_pincache_clear: called
0x7fa0755b7740 21:02:06.980 [opensc-pkcs11] misc.c:61:sc_to_cryptoki_error_common: libopensc return value: 0 (Success)
0x7fa0755b7740 21:02:06.980 [opensc-pkcs11] card.c:346:sc_disconnect_card: called
0x7fa0755b7740 21:02:06.980 [opensc-pkcs11] card-dnie.c:888:dnie_finish: called
0x7fa0755b7740 21:02:06.980 [opensc-pkcs11] cwa14890.c:1035:cwa_create_secure_channel: called
0x7fa0755b7740 21:02:06.980 [opensc-pkcs11] cwa14890.c:1041:cwa_create_secure_channel: Setting CWA SM status to none
0x7fa0755b7740 21:02:06.980 [opensc-pkcs11] cwa14890.c:1042:cwa_create_secure_channel: returning with: 0 (Success)
0x7fa0755b7740 21:02:06.980 [opensc-pkcs11] card-dnie.c:894:dnie_finish: returning with: 0 (Success)
0x7fa0755b7740 21:02:07.018 [opensc-pkcs11] reader-pcsc.c:597:pcsc_disconnect: Generic USB2.0-CRW [Smart Card Reader Interface] (20070818000000000) 00 00:SCardDisconnect returned: 0x00000000
0x7fa0755b7740 21:02:07.018 [opensc-pkcs11] card.c:368:sc_disconnect_card: returning with: 0 (Success)
0x7fa0755b7740 21:02:07.018 [opensc-pkcs11] ctx.c:870:sc_release_context: called
0x7fa0755b7740 21:02:07.018 [opensc-pkcs11] reader-pcsc.c:896:pcsc_finish: called
error: PKCS11 function C_Sign failed: rv = CKR_ARGUMENTS_BAD (0x7)
Aborting.
```

Sorry I omit so much, but i don't wish to disclose too much information.
Maybe I can help if you tell me what should I look for.

**Comment by frankmorgner** [comment]
Thanks, I consider that a good thing.

I've submitted a fix for the second problem...

**Comment by alvaroaalvarez** [comment]
Thanks, it's working now.
Still get errors, maybe it's some feature not available, but I can now use the card to log in to websites

```
[alvaro@dell8700 ~]$ /opt/opensc-dnie/bin/pkcs11-tool -t -l
Using slot 0 with a present token (0x0)
Logging in to "PIN1 (DNI electrónico)".
Please enter User PIN: 
C_SeedRandom() and C_GenerateRandom():
  seeding (C_SeedRandom) not supported
  ERR: C_GenerateRandom(buf1,0) failed: CKR_ARGUMENTS_BAD (0x7)
Digests:
  all 4 digest functions seem to work
  MD5: OK
  SHA-1: OK
  RIPEMD160: OK
Signatures (currently only for RSA)
  testing key 0 (KprivAutenticacion) 
  all 4 signature functions seem to work
  testing signature mechanisms:
    RSA-PKCS: OK
    SHA1-RSA-PKCS: OK
    MD5-RSA-PKCS: OK
    RIPEMD160-RSA-PKCS: OK
    SHA256-RSA-PKCS: OK
  testing key 1 (1920 bits, label=KprivFirmaDigital) with 1 signature mechanism
    RSA-PKCS: OK
Verify (currently only for RSA)
  testing key 0 (KprivAutenticacion)
    RSA-PKCS: OK
    SHA1-RSA-PKCS: OK
    MD5-RSA-PKCS: OK
    RIPEMD160-RSA-PKCS: OK
  testing key 1 (KprivFirmaDigital) with 1 mechanism
    RSA-PKCS: OK
Unwrap: not implemented
Decryption (currently only for RSA)
  testing key 0 (KprivAutenticacion)  -- can't be used to decrypt, skipping
  testing key 1 (KprivFirmaDigital)  -- can't be used to decrypt, skipping
1 errors
```

BTW, I changed a line in the code:
`1706: 	if ((int)outlen<result_resplen)`

**Comment by frankmorgner** [comment]
Sounds good.

If you send a log of the errors, I'll have a look

**Comment by alvaroaalvarez** [comment]
Hope this helps

```
0x7eff61f03880 18:50:30.763 [opensc-pkcs11] apdu.c:390:sc_single_transmit: returning with: 0 (Success)
0x7eff61f03880 18:50:30.763 [opensc-pkcs11] apdu.c:543:sc_transmit: returning with: 0 (Success)
0x7eff61f03880 18:50:30.763 [opensc-pkcs11] card.c:459:sc_unlock: called
0x7eff61f03880 18:50:30.763 [opensc-pkcs11] card-dnie.c:1809:dnie_check_sw: called
0x7eff61f03880 18:50:30.763 [opensc-pkcs11] iso7816.c:124:iso7816_check_sw: File not found
0x7eff61f03880 18:50:30.763 [opensc-pkcs11] card-dnie.c:1821:dnie_check_sw: returning with: -1201 (File not found)
0x7eff61f03880 18:50:30.763 [opensc-pkcs11] card-dnie.c:1184:dnie_compose_and_send_apdu: SelectFile() check_sw failed: -1201 (File not found)
0x7eff61f03880 18:50:30.763 [opensc-pkcs11] card.c:459:sc_unlock: called
0x7eff61f03880 18:50:30.763 [opensc-pkcs11] reader-pcsc.c:662:pcsc_unlock: called
0x7eff61f03880 18:50:30.773 [opensc-pkcs11] card-dnie.c:1294:dnie_select_file: select_file(PATH) failed: -1201 (File not found)
0x7eff61f03880 18:50:30.773 [opensc-pkcs11] card.c:770:sc_select_file: 'SELECT' error: -1201 (File not found)
0x7eff61f03880 18:50:30.773 [opensc-pkcs11] dir.c:171:sc_enum_apps: Cannot select EF.DIR file: -1201 (File not found)
0x7eff61f03880 18:50:30.773 [opensc-pkcs11] slot.c:304:card_detect: Generic USB2.0-CRW [Smart Card Reader Interface] (20070818000000000) 00 00: Detecting Framework. 0 on-card applications
0x7eff61f03880 18:50:30.773 [opensc-pkcs11] slot.c:305:card_detect: Generic USB2.0-CRW [Smart Card Reader Interface] (20070818000000000) 00 00: generic application <none>
0x7eff61f03880 18:50:30.773 [opensc-pkcs11] slot.c:317:card_detect: Generic USB2.0-CRW [Smart Card Reader Interface] (20070818000000000) 00 00: Detected framework 0. Creating tokens.
...
0x7eff61f03880 18:50:36.402 [opensc-pkcs11] cwa14890.c:1822:cwa_decode_response: returning with: 0 (Success)
0x7eff61f03880 18:50:36.402 [opensc-pkcs11] card-dnie.c:774:dnie_sm_free_wrapped_apdu: returning with: 0 (Success)
0x7eff61f03880 18:50:36.402 [opensc-pkcs11] sm.c:173:sc_sm_single_transmit: returning with: 0 (Success)
0x7eff61f03880 18:50:36.402 [opensc-pkcs11] apdu.c:382:sc_single_transmit: returning with: 0 (Success)
0x7eff61f03880 18:50:36.402 [opensc-pkcs11] apdu.c:543:sc_transmit: returning with: 0 (Success)
0x7eff61f03880 18:50:36.402 [opensc-pkcs11] card.c:459:sc_unlock: called
0x7eff61f03880 18:50:36.402 [opensc-pkcs11] reader-pcsc.c:662:pcsc_unlock: called
0x7eff61f03880 18:50:36.403 [opensc-pkcs11] card-dnie.c:1385:dnie_get_challenge: returning with: 0 (Success)
0x7eff61f03880 18:50:36.403 [opensc-pkcs11] card.c:825:sc_get_challenge: returning with: 0 (Success)
0x7eff61f03880 18:50:36.403 [opensc-pkcs11] misc.c:61:sc_to_cryptoki_error_common: libopensc return value: 0 (Success)
0x7eff61f03880 18:50:36.403 [opensc-pkcs11] card.c:819:sc_get_challenge: called
0x7eff61f03880 18:50:36.403 [opensc-pkcs11] card-dnie.c:1350:dnie_get_challenge: called
0x7eff61f03880 18:50:36.403 [opensc-pkcs11] card-dnie.c:1385:dnie_get_challenge: returning with: -1300 (Invalid arguments)
0x7eff61f03880 18:50:36.403 [opensc-pkcs11] card.c:825:sc_get_challenge: returning with: -1300 (Invalid arguments)
0x7eff61f03880 18:50:36.403 [opensc-pkcs11] misc.c:61:sc_to_cryptoki_error_common: libopensc return value: -1300 (Invalid arguments)
  ERR: C_GenerateRandom(buf1,0) failed: CKR_ARGUMENTS_BAD (0x7)
0x7eff61f03880 18:50:36.403 [opensc-pkcs11] pkcs11-session.c:203:C_GetSessionInfo: C_GetSessionInfo(hSession:0xd21740)
0x7eff61f03880 18:50:36.403 [opensc-pkcs11] pkcs11-session.c:211:C_GetSessionInfo: C_GetSessionInfo(slot:0x0)
0x7eff61f03880 18:50:36.404 [opensc-pkcs11] pkcs15-pin.c:689:sc_pkcs15_get_pin_info: called
0x7eff61f03880 18:50:36.404 [opensc-pkcs11] card.c:407:sc_lock: called
0x7eff61f03880 18:50:36.404 [opensc-pkcs11] reader-pcsc.c:612:pcsc_lock: called
0x7eff61f03880 18:50:36.404 [opensc-pkcs11] card.c:449:sc_lock: returning with: 0 (Success)
0x7eff61f03880 18:50:36.404 [opensc-pkcs11] sec.c:170:sc_pin_cmd: called
0x7eff61f03880 18:50:36.404 [opensc-pkcs11] card-dnie.c:2237:dnie_pin_cmd: called
0x7eff61f03880 18:50:36.404 [opensc-pkcs11] card-dnie.c:1845:dnie_card_ctl: called
0x7eff61f03880 18:50:36.404 [opensc-pkcs11] card-dnie.c:1865:dnie_card_ctl: returning with: 0 (Success)
0x7eff61f03880 18:50:36.404 [opensc-pkcs11] card-dnie.c:2281:dnie_pin_cmd: returning with: -1408 (Not supported)
0x7eff61f03880 18:50:36.404 [opensc-pkcs11] sec.c:217:sc_pin_cmd: returning with: -1408 (Not supported)
0x7eff61f03880 18:50:36.404 [opensc-pkcs11] card.c:459:sc_unlock: called
0x7eff61f03880 18:50:36.404 [opensc-pkcs11] reader-pcsc.c:662:pcsc_unlock: called
0x7eff61f03880 18:50:36.406 [opensc-pkcs11] pkcs15-pin.c:717:sc_pkcs15_get_pin_info: returning with: -1408 (Not supported)
0x7eff61f03880 18:50:36.406 [opensc-pkcs11] pkcs11-session.c:229:C_GetSessionInfo: C_GetSessionInfo(0xd21740) = CKR_OK
...
0x7eff61f03880 18:50:40.134 [opensc-pkcs11] apdu.c:390:sc_single_transmit: returning with: 0 (Success)
0x7eff61f03880 18:50:40.134 [opensc-pkcs11] apdu.c:543:sc_transmit: returning with: 0 (Success)
0x7eff61f03880 18:50:40.134 [opensc-pkcs11] card.c:459:sc_unlock: called
0x7eff61f03880 18:50:40.134 [opensc-pkcs11] reader-pcsc.c:662:pcsc_unlock: called
0x7eff61f03880 18:50:40.144 [opensc-pkcs11] card-dnie.c:1809:dnie_check_sw: called
0x7eff61f03880 18:50:40.144 [opensc-pkcs11] card-dnie.c:1814:dnie_check_sw: Expected SM Data Object missing
0x7eff61f03880 18:50:40.144 [opensc-pkcs11] card-dnie.c:1184:dnie_compose_and_send_apdu: SelectFile() check_sw failed: -1600 (Generic Secure Messaging error)
0x7eff61f03880 18:50:40.144 [opensc-pkcs11] card-dnie.c:1423:dnie_logout: returning with: 0 (Success)
0x7eff61f03880 18:50:40.144 [opensc-pkcs11] pkcs11-global.c:315:C_Finalize: C_Finalize()
0x7eff61f03880 18:50:40.144 [opensc-pkcs11] ctx.c:873:sc_cancel: called
0x7eff61f03880 18:50:40.144 [opensc-pkcs11] reader-pcsc.c:712:pcsc_cancel: called
0x7eff61f03880 18:50:40.144 [opensc-pkcs11] slot.c:195:card_removed: Generic USB2.0-CRW [Smart Card Reader Interface] (20070818000000000) 00 00: card removed
0x7eff61f03880 18:50:40.144 [opensc-pkcs11] slot.c:474:slot_token_removed: slot_token_removed(0x0)
0x7eff61f03880 18:50:40.144 [opensc-pkcs11] pkcs11-session.c:140:sc_pkcs11_close_all_sessions: real C_CloseAllSessions(0x0) 0
0x7eff61f03880 18:50:40.144 [opensc-pkcs11] framework-pkcs15.c:1498:pkcs15_release_token: pkcs15_release_token() not implemented
0x7eff61f03880 18:50:40.144 [opensc-pkcs11] slot.c:474:slot_token_removed: slot_token_removed(0x1)
0x7eff61f03880 18:50:40.144 [opensc-pkcs11] pkcs11-session.c:140:sc_pkcs11_close_all_sessions: real C_CloseAllSessions(0x1) 0
0x7eff61f03880 18:50:40.144 [opensc-pkcs11] slot.c:474:slot_token_removed: slot_token_removed(0x2)
0x7eff61f03880 18:50:40.144 [opensc-pkcs11] pkcs11-session.c:140:sc_pkcs11_close_all_sessions: real C_CloseAllSessions(0x2) 0
0x7eff61f03880 18:50:40.144 [opensc-pkcs11] slot.c:474:slot_token_removed: slot_token_removed(0x3)
0x7eff61f03880 18:50:40.144 [opensc-pkcs11] pkcs11-session.c:140:sc_pkcs11_close_all_sessions: real C_CloseAllSessions(0x3) 0
0x7eff61f03880 18:50:40.144 [opensc-pkcs11] sc.c:275:sc_detect_card_presence: called
0x7eff61f03880 18:50:40.144 [opensc-pkcs11] reader-pcsc.c:411:pcsc_detect_card_presence: called
0x7eff61f03880 18:50:40.144 [opensc-pkcs11] reader-pcsc.c:319:refresh_attributes: Generic USB2.0-CRW [Smart Card Reader Interface] (20070818000000000) 00 00 check
0x7eff61f03880 18:50:40.145 [opensc-pkcs11] reader-pcsc.c:339:refresh_attributes: returning with: 0 (Success)
0x7eff61f03880 18:50:40.145 [opensc-pkcs11] reader-pcsc.c:416:pcsc_detect_card_presence: returning with: 1
0x7eff61f03880 18:50:40.145 [opensc-pkcs11] sc.c:280:sc_detect_card_presence: returning with: 1
0x7eff61f03880 18:50:40.145 [opensc-pkcs11] pkcs15.c:1314:sc_pkcs15_unbind: called
0x7eff61f03880 18:50:40.145 [opensc-pkcs11] pkcs15-pin.c:838:sc_pkcs15_pincache_clear: called
0x7eff61f03880 18:50:40.145 [opensc-pkcs11] misc.c:61:sc_to_cryptoki_error_common: libopensc return value: 0 (Success)
0x7eff61f03880 18:50:40.145 [opensc-pkcs11] card.c:346:sc_disconnect_card: called
0x7eff61f03880 18:50:40.145 [opensc-pkcs11] card-dnie.c:886:dnie_finish: called
0x7eff61f03880 18:50:40.145 [opensc-pkcs11] cwa14890.c:1035:cwa_create_secure_channel: called
0x7eff61f03880 18:50:40.145 [opensc-pkcs11] cwa14890.c:1041:cwa_create_secure_channel: Setting CWA SM status to none
0x7eff61f03880 18:50:40.145 [opensc-pkcs11] cwa14890.c:1042:cwa_create_secure_channel: returning with: 0 (Success)
0x7eff61f03880 18:50:40.145 [opensc-pkcs11] card-dnie.c:892:dnie_finish: returning with: 0 (Success)
0x7eff61f03880 18:50:40.183 [opensc-pkcs11] reader-pcsc.c:597:pcsc_disconnect: Generic USB2.0-CRW [Smart Card Reader Interface] (20070818000000000) 00 00:SCardDisconnect returned: 0x00000000
0x7eff61f03880 18:50:40.183 [opensc-pkcs11] card.c:368:sc_disconnect_card: returning with: 0 (Success)
0x7eff61f03880 18:50:40.183 [opensc-pkcs11] ctx.c:898:sc_release_context: called
0x7eff61f03880 18:50:40.183 [opensc-pkcs11] reader-pcsc.c:896:pcsc_finish: called
```

**Comment by frankmorgner** [comment]
The problem with `C_GenerateRandom` should be fixed with #1249.

**Comment by howl** [comment]
Using OpenSC with a new DNI 3.0 from December 2017 doesn't work, leave a blank page in Firefox and with Chromium it throws ERR_SSL_CLIENT_AUTH_SIGNATURE_FAILED.

Using the official library works. Differences I could appreciate are that listing PKCS modules with the official one always make the reader to read the card, and the official one reports the HW FW version from the card and not from the reader where the OpenSC reports the reader HW FW versions and not the card HS FW versions.

Could this be related?

I had certificates from December 2017 but I renewed them in April 2018 to discard certificate issues despite the certificates with troubles are until November 2017.
